atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/ConsistentOrderingObjectMapper.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.sweep.asts.bucketingthings;
+
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.MapperFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.conjure.java.serialization.ObjectMappers;
+
+/**
+ * The object mapper present in this class is configured to ensure that, given the same input object, the serialised
+ * form is _always_ syntactically the same. This is important when performing a CAS operation from one present value
+ * to another, where the value is a serialised object. If the serialised form changes, the CAS operation will fail.
+ *
+ * Moreover, in the presence of new Optional fields that are empty, the serialised form will be the same as if the
+ * field was not present at all. This is important for backwards compatibility - when adding new fields to the
+ * object, must still be able to retrieve the old serialised format of the object for CAS operations over old values.
```

Comment:
We probably want to write down some limits on what this can do. If I'm not mistaken,

```
interface InterestingTypes {
  Set<Comparable> comparables(); // maybe we should support this one
  Set<Uncomparable> uncomparables(); // we almost certainly shouldn't support this!
  Stream<Object> unorderedStream(); // good luck
}
```

wouldn't be consistently handled with the class as written. I think we should try to handle `comparables` (if not we say that the user is responsible), but we should explicitly say `uncomparables` and streams are undefined behaviour. Maybe even for `comparables` we might not want to handle it if it's too hard.

Reply:
It's certainly possible: https://stackoverflow.com/questions/57815178/how-can-i-get-jackson-to-serialize-an-unsorted-set-in-sorted-order#:~:text=But%20if%20you%20want%20to,be%20checked%20in%20custom%20implementation.

I'm going to omit given we don't need this, but it's not terribly hard to do for the easy cases (and we can just note we don't handle the other cases)